contrib/udfs/src/main/java/org/apache/drill/exec/udfs/NearestDateUtils.java

```diff
@@ -0,0 +1,142 @@
+    int second = d.getSecond();
+
+    try {
+      switch (TimeInterval.valueOf(interval.toUpperCase())) {
+        case YEAR:
+          newDate = java.time.LocalDateTime.of(year, 1, 1, 0, 0, 0);
+          break;
+        case QUARTER:
+          newDate = java.time.LocalDateTime.of(year, (month / 3) * 3 + 1, 1, 0, 0, 0);
+          break;
+        case MONTH:
+          newDate = java.time.LocalDateTime.of(year, month, 1, 0, 0, 0);
+          break;
+        case WEEK_SUNDAY:
+          newDate = newDate.with(java.time.temporal.TemporalAdjusters.previousOrSame(java.time.DayOfWeek.SUNDAY))
+                  .truncatedTo(java.time.temporal.ChronoUnit.DAYS);
+          break;
+        case WEEK_MONDAY:
+          newDate = newDate.with(java.time.temporal.TemporalAdjusters.previousOrSame(java.time.DayOfWeek.MONDAY))
+                  .truncatedTo(java.time.temporal.ChronoUnit.DAYS);
+          break;
+        case DAY:
+          newDate = java.time.LocalDateTime.of(year, month, day, 0, 0, 0);
+          break;
+        case HOUR:
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, 0, 0);
+          break;
+        case HALF_HOUR:
+          if (minute >= 30) {
+            minute = 30;
+          } else {
+            minute = 0;
+          }
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, minute, 0);
+          break;
+        case QUARTER_HOUR:
+          if (minute >= 45) {
+            minute = 45;
+          } else if (minute >= 30) {
+            minute = 30;
+          } else if (minute >= 15) {
+            minute = 15;
+          } else {
+            minute = 0;
+          }
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, minute, 0);
+          break;
+        case MINUTE:
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, minute, 0);
+          break;
+        case HALF_MINUTE:
+          if (second >= 30) {
+            second = 30;
+          } else {
+            second = 0;
+          }
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, minute, second);
+          break;
+        case QUARTER_MINUTE:
+          if (second >= 45) {
+            second = 45;
+          } else if (second >= 30) {
+            second = 30;
+          } else if (second >= 15) {
+            second = 15;
+          } else {
+            second = 0;
+          }
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, minute, second);
+          break;
+        case SECOND:
+          newDate = java.time.LocalDateTime.of(year, month, day, hour, minute, second);
+          break;
+        default:
+          throw  org.apache.drill.common.exceptions.UserException.validationError()
+                  .message(interval + " is not a valid time statement.")
+                  .build(logger);
+      }
+    } catch (Exception e) {
+      throw  org.apache.drill.common.exceptions.UserException.validationError()
```

Comment:
1. Catch only specific exception.
2. Since this class is static, i.e. utility, I think it's a bit awkward that it throws user exception, please change it to DrillRuntimeException.